lib/utils/ignore.ts

```diff
@@ -200,6 +200,41 @@ export interface FileStats {
 	stats: Stats;
 }
 
+/**
+ * Create a list of files for the filesystem subtree rooted at
+ * projectDir, excluding entries for directories themselves.
+ * @param projectDir Source directory (root of subtree to be listed)
+ */
+async function listFiles(projectDir: string): Promise<string[]> {
+	const dirs: string[] = [];
+	const files: string[] = [];
+	dirs.push(projectDir);
+	async function walk(currentDirs: string[]): Promise<string[]> {
+		if (!currentDirs.length) {
+			return files;
+		}
+
+		const foundDirs: string[] = [];
+
+		await Promise.all(
+			currentDirs.map(async (dir) => {
+				const _files = await fs.readdir(dir, { withFileTypes: true });
+				for (const entry of _files) {
+					const fpath = `${dir}/${entry.name}`;
+					if (entry.isDirectory()) {
+						foundDirs.push(fpath);
+						dirs.push(fpath);
+					} else {
+						files.push(fpath);
```

Comment:
To be confirmed:
I assume (didn't test) that `fpath` here could be a symbolic link or special file (like socket, block or character devices, named pipes, etc). If a symbolic link, it is not followed (resolved) here, and as currently implemented, I think the dockerignore tests are made against the unresolved symlinks rather than against the files they point to.

I believe the `klaw` implementation followed (resolved) symlinks, and then discarded entries that weren't a regular file, all _before_ the dockerignore tests were executed on the entries:
```ts
				if (!stats.isFile()) {
					return;
				}
```

Unless there are reasons not to (such as noticeably hurting performance), I would favour maintaining existing behaviour (otherwise, strictly, it would be a breaking change). Existing tests may or may not be catching all these scenarios and yes, it would be nice to improve the test cases. :-)

Reply:
@pdcastro I don't think that `klaw` resolves symlinks, if it did the tests that we have for symlinked files would fail. Could you create a test case that runs the scenario you are describing?

Reply:
I've added a test case for Unix domain socket files in #2384, rebased this PR and now the build is failing - sorry! 😀 
See also this comment: https://github.com/balena-io/balena-cli/pull/2376#issuecomment-975017791

Re symlinks, I have not (yet) determined whether or not `klaw` resolves symlinks at this point, but potentially it does not matter\* because symlinks will eventually be resolved in [readFile(fileStats.filePath)](https://github.com/balena-io/balena-cli/blob/v12.52.1/lib/utils/compose_ts.ts#L812) if they are not resolved here.

\* Except for pre-existing bug #2383, which I propose be fixed in some other PR to avoid delaying this one.